python/ray/experimental/dag/vis_utils.py

```diff
@@ -0,0 +1,112 @@
+from ray.experimental.dag.dag_node import DAGNode
+from ray.serve.pipeline.generate import DeploymentNameGenerator
+
+import os
+import tempfile
+
+
+def check_pydot_and_graphviz():
+    """Check if pydot and graphviz are installed.
+
+    pydot and graphviz are required for plotting. We check this
+    during runtime rather than adding them to Ray dependencies.
+
+    """
+    try:
+        import pydot
+    except ImportError:
+        raise ImportError(
+            "pydot is required to plot DAG, " "install it with `pip install pydot`."
+        )
+    try:
+        pydot.Dot.create(pydot.Dot())
+    except (OSError, pydot.InvocationException):
+        raise ImportError(
+            "graphviz is required to plot DAG, "
+            "download it from https://graphviz.gitlab.io/download/"
+        )
+
+
+def get_nodes_and_edges(dag: DAGNode):
+    """Get all unique nodes and edges in the DAG.
+
+    A basic dfs with memorization to get all unique nodes
+    and edges in the DAG.
+    Unique nodes will be used to generate unique names,
+    while edges will be used to construct the graph.
+    """
+
+    def _dfs(node, res, visited):
+        if node not in visited:
+            visited.append(node)
+            for child_node in node._get_all_child_nodes():
+                res.append((child_node, node))
+                _dfs(child_node, res, visited)
+
+    edges = []
+    nodes = []
```

Comment:
this could be a set to make the dfs visited check O(1) ? Feels like leetcoding again :D